File: bms/go/geo/geo_s1_ChordAngleFromAngle/geo_s1_ChordAngleFromAngle.go
package test

import (
	"math"
)

var ()

const (
	geo_s1_NegativeChordAngle = geo_s1_ChordAngle(-1)
)

func geo_s1_ChordAngleFromAngle(a geo_s1_Angle) geo_s1_ChordAngle {
	if a < 0 {
		return geo_s1_NegativeChordAngle
	}
	if a.isInf() {
		return geo_s1_InfChordAngle()
	}
	l := 2 * math.Sin(0.5*math.Min(math.Pi, a.Radians()))
	return geo_s1_ChordAngle(l * l)
}

func (a geo_s1_Angle) isInf() bool {
	return math.IsInf(float64(a), 0)
}

func geo_s1_InfChordAngle() geo_s1_ChordAngle {
	return geo_s1_ChordAngle(math.Inf(1))
}

func (a geo_s1_Angle) Radians() float64 { return float64(a) }

type geo_s1_Angle float64

type geo_s1_ChordAngle float64

File: translations_only/success/go/geo/geo_s1_ChordAngleFromAngle/gpt4/BaseRepair/geo_s1_ChordAngleFromAngle.rs

use std::f64::consts::PI;

const GEO_S1_NEGATIVE_CHORD_ANGLE: GeoS1ChordAngle = GeoS1ChordAngle(-1.0);

fn geo_s1_chord_angle_from_angle(a: GeoS1Angle) -> GeoS1ChordAngle {
    if a.0 < 0.0 {
        return GEO_S1_NEGATIVE_CHORD_ANGLE;
    }
    if a.is_inf() {
        return geo_s1_inf_chord_angle();
    }
    let l = 2.0 * (0.5 * a.radians()).min(PI / 2.0).sin();
    GeoS1ChordAngle(l * l)
}

fn geo_s1_inf_chord_angle() -> GeoS1ChordAngle {
    GeoS1ChordAngle(f64::INFINITY)
}

#[derive(Copy, Clone)]
struct GeoS1Angle(f64);

impl GeoS1Angle {
    fn is_inf(self) -> bool {
        self.0.is_infinite()
    }

    fn radians(self) -> f64 {
        self.0
    }
}

#[derive(Copy, Clone)]
struct GeoS1ChordAngle(f64);
